### ISSUE TITLE
Need better infrastructure for detecting unintentional breaking changes to analyzer

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#42160]
The analyzer package has a lot of clients that are widely used in open source, and several of them import code from `package:analyzer/src/...`.  As a result, we occasionally make changes to the analyzer that don't break the officially-supported analyzer API, but in practice break an important client.  To make sure that we don't publish a version of the analyzer that would cause a widespread ecosystem breakage, we need a way to detect these "de facto" breaking changes early so we can revert them.

In the past, our best signal for detecting these breaking changes has been to notice them when we roll the analyzer into the Google internal codebase, however that roll process has been becoming more atomic, so that signal is becoming less reliable.  It would be really nice to have a strong signal for detecting such breakages that *doesn't* rely on any internal infrastructure.

**Comment by bwilkerson** [comment]
If we believed that is was possible for the analyzer to publish an API that clients wouldn't need to circumvent, then it seems to me that we'd have enough information to know what is and is not a breaking change. I don't know whether that's a reasonable scenario, but might be worth considering.

**Comment by stereotype441** [comment]
See #42163 for a specific example of the kind of unintentional breaking change that I hope we could avoid with better testing

**Comment by stereotype441** [comment]
One possible way we could address this using testing would be to set up a bot with a list of "known important" packages.  For each published version of each of those packages, it does the following:
- Check whether the package's analyzer dependency includes the bleeding edge version of the analyzer.  If it doesn't, skip it.
- Download the package into a temporary directory.
- Modify the package's pubspec, adding a dependency override to point to the bleeding edge analyzer.
- Run `pub get`
- Run some sanity checks (e.g. analyze the package, run `pub run test` if the package supports it)

It might be too slow a process to run on every single commit, but we could certainly run it on a regular basis.

**Comment by stereotype441** [comment]
> If we believed that is was possible for the analyzer to publish an API that clients wouldn't need to circumvent, then it seems to me that we'd have enough information to know what is and is not a breaking change. I don't know whether that's a reasonable scenario, but might be worth considering.

Agreed, if we had confidence that were feasible, it would sure be nice.  My personal feeling is that it isn't feasible, based on several observations:
- We've been talking about trying to get the analyzer's API into a "1.0" state for years and we still seem pretty far from achieving it (at least measured by the number of client packages that import from lib/src).  Granted, a 1.0 state is not necessarily the same thing as "no clients have to import from lib/src" but I think it's reasonable to assume that the effort to achieve the two goals is in the same ballpack of difficulty.
- In our past efforts to get to a 1.0 state and get clients to stop importing from lib/src, it hasn't simply been a matter of moving the analyzer APIs that are actually used out of lib/src; we've often found tht the client packages are using the analyzer in the "wrong" way (or, at least, in a way that we think is a bad idea and don't want to support as an official part of the public API), and that it takes a lot of coding work to get the clients to use the analyzer in the "right" way.
- Even if we can get into a state where clients aren't importing from lib/src, it's possible that a client will be relying on a behavior that doesn't obviously manifest in the analyzer's public API signatures (e.g. expecting a specific error message when handling invalid code).  It would be nice if we had a way to notice breakages caused by these assumptions too.

That's not to say that we shouldn't continue trying to move toward a 1.0 state and to reduce the amount of imports of `lib/src`; those are good ideals and we should keep working toward them.  But IMHO it will take a long time to reach them (if ever), and I think a testing based solution would be really helpful in the meantime.
